File: Step 04 Binary Search/Step 4.1 Learning BS on 1D Array/LeetCode/Floor_Ceil in sorted array.java
import java.util.*;
import java.io.*;

public class Solution {
    static int getFloor(int[] arr, int n, int x) {
        int low = 0;
        int high = n - 1;
        // int ans = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x) return arr[mid];
            else if(arr[mid] < x) low = mid + 1;
            else high = mid - 1;
        }
        return high == -1 ? -1 : arr[high];
    }

    static int getCeil(int[] arr, int n, int x) {
        int low = 0;
        int high = n - 1;
        // int ans = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x) return arr[mid];
            else if(arr[mid] < x) low = mid + 1;
            else high = mid - 1;
        }
        return low == n ? -1 : arr[low];
    }

    public static int[] getFloorAndCeil(int[] arr, int n, int x) {
        int[] res = new int[2];
        res[0] = getCeil(arr, n, x);
        res[1] = getFloor(arr, n, x);
        return res;
    }
}

File: Step 04 Binary Search/Step 4.1 Learning BS on 1D Array/Floor_Ceil in sorted array/Floor_Ceil in sorted array.cpp
// User code template

class Solution {
  public:
  
  int getFloor(vector<int> &arr, int x) {
      int n = arr.size();
      int low = 0;
       int high = n - 1;
       int floorEle = -1;
       while (low <= high) {
           int mid = low + (high - low) / 2;
           if (arr[mid] == x) return arr[mid];
           else if (arr[mid] < x) {
               floorEle = arr[mid];
               low = mid + 1;
           } else high = mid - 1;
       }
       return floorEle;
  }
  
  int getCeil(vector<int> &arr, int x) {
      int n = arr.size();
      int low = 0;
       int high = n - 1;
       int ceilEle = -1;
       while (low <= high) {
           int mid = low + (high - low) / 2;
           if (arr[mid] == x) return arr[mid];
           else if (arr[mid] < x) {
               low = mid + 1;
           } else {
               ceilEle = arr[mid];
               high = mid - 1;
               
           }
       }
       return ceilEle;
  }
  
    vector<int> getFloorAndCeil(int x, vector<int> &arr) {
        // code here
        sort(arr.begin(), arr.end());
        vector<int> p(2);
        p[0] = getFloor(arr, x);
        p[1] = getCeil(arr, x);
        return p;
    }
};